app/controllers/planning_applications_controller.rb

```diff
@@ -120,7 +123,13 @@ def invalidate
       @planning_application.invalidate!
       audit("invalidated")
       invalidation_notice_mail
-      @planning_application.unsent_validation_requests.each { |request| request.update!(notified_at: Time.zone.now) }
+      request_names = []
+      @planning_application.unsent_validation_requests.each do |request|
```

Comment:
can you make this a callback on the model's state transition instead? Something a bit like:

```diff
modified   app/models/planning_application.rb
@@ -64,7 +64,10 @@ class PlanningApplication < ApplicationRecord
     end
 
     event :invalidate do
-      transitions from: %i[not_started invalidated], to: :invalidated, guard: :validation_requests_open?
+      transitions from: %i[not_started invalidated],
+                  to: :invalidated,
+                  guard: :validation_requests_open?,
+                  after: :send_validation_requests
     end
 
     event :determine do
@@ -104,6 +107,14 @@ class PlanningApplication < ApplicationRecord
     "#{applicant_first_name} #{applicant_last_name}"
   end
 
+  def send_validation_requests
+    unsent_validation_requests.each do |request|
+      request.update!(notified_at: Time.zone.now)
+
+      audit("validation_request_sent", nil, request)
+    end
+  end
+
   def timestamp_status_change
     update("#{aasm.to_state}_at": Time.zone.now)
   end
```

Reply:
I can try for sure! Will give it a spin - Silly question but, we are ok to call mailer methods from the model? Will it know the request.host still? since we redefine on the controller this method:
``` 
def invalidation_notice_mail
    PlanningApplicationMailer.invalidation_notice_mail(
      @planning_application,
      request.host,
    ).deliver_now
  end
```

Reply:
good question; the mailing should remain in the controller I think, it's in charge of orchestrating that stuff: invalidate application + mail about it. The whole "when an application is invalidated all requests should be sent out" belongs to the model. I was actually in Show&Tell prep yesterday and didn't realise that was how the flow worked, so I think a

```ruby
    event :invalidate do
      transitions from: %i[not_started invalidated],
                  to: :invalidated,
                  guard: :validation_requests_open?,
                  after: :send_all_pending_validation_requests # or before, not sure
    end
```

will help a lot in that respect.